### ISSUE TITLE
False positive for F821

### ISSUE THREAD

**Issue by olliemath** [charliermarsh/ruff#780]
It seems that using `dict` as a parameter name in a function (which I'd consider bad style) breaks ruff's parsing in unexpected ways. In particular the following gives `6:31: F821 Undefined name 'name'`, when it shouldn't. 

```python
def unimportant(name):
    pass


def dang(dict):
    return unimportant(name=dict["name"])
```

Here is my pyproject.toml if that's needed to reproduce:
```toml
[tool.ruff]
ignore = [
    'B007',
    'E741',
    'E501',
]
line-length = 88
select = [
    'B',
    'E',
    'F',
    'W',
]
```

**Comment by charliermarsh** [comment]
Thanks! Yeah, definitely a legit bug. We need to verify that `dict` corresponds to the built-in and hasn't been shadowed.

**Comment by charliermarsh** [comment]
Appreciate the clear test case.

**Comment by charliermarsh** [comment]
(I'll fix this tomorrow most likely, probably not tonight.)

**Comment by charliermarsh** [comment]
Tangentially, I guess I'd argue that Flake8 erroneously _doesn't_ catch this:

```py
# Flake8 allows this
x = dict["foo"]

# Flake8 raises F821 here
x: dict["foo"]
```

Both should raise an undefined error on `foo`.

**Comment by charliermarsh** [comment]
The reason I mention it is because both Flake8 and Ruff accidentally raise F821 here:

```py
def f(dict):
    x: dict["foo"] = 1
```

**Comment by charliermarsh** [comment]
But I will fix this in Ruff :)

**Comment by olliemath** [comment]
Thanks for the quick response! Yeah interesting that flask has that issue too - I guess it's a bit more niche, so as-yet undiscovered (or unfixed) by them.

**Comment by olliemath** [comment]
Well, I've also let the pyflakes team know about this edge case https://github.com/PyCQA/pyflakes/issues/738 I assume they'll want to fix it!

**Comment by olliemath** [comment]
> The reason I mention it is because both Flake8 and Ruff accidentally raise F821 here:
> 
> ```python
> def f(dict):
>     x: dict["foo"] = 1
> ```

So pyflakes adopt the attitude of "this is a bad type annotation - won't fix". In fact this also gives F821 with flake8 but not ruff:
```def f(noshadow):
    x: noshadow["foo"] = 1
```

I think it's kind of ambiguous if it should give F821, since annotations can be used for other things than types (but rarely are). To be honest, I would let mypy and co sort out the correctness of my type annotations.

Either way, having consistent behavior in ruff with or without shadowing is good (whether or not that differs from flake8 slightly).

**Comment by charliermarsh** [comment]
I'm behind on this, it'll come soon :)

**Comment by JonathanPlasse** [comment]
@charliermarsh, I could work on it if it is ok with you.

**Comment by charliermarsh** [comment]
That'd be great @JonathanPlasse! I know it's relevant to the `sys.exit` stuff too.
